### ISSUE TITLE
Suggestion Feature. 'compose' method with deep extend.

### ISSUE THREAD

**Issue by raabbajam** [ericelliott/stampit#40]
Can we use deep recursive copy / extend on 'compose' method?
Right now it doesn't. Can't we do this?
```javascript
var Validator = require('../libs/Validator');
// Validator has state.rules.flight = { ori: 'required|size:3', dst: 'required|size:3' }
// I want to create another factory with 
//  rules.flight = { ori: 'size:3', dst: 'required|size:3', rute: 'required' }; 
//  with 'ori' now is not required and 'rute' is required
describe('#compose with other states', function() {
		var _states = {
			rules: {
				flight: {
					ori: 'size:3',
                    rute: 'required',
				}
			}
		};
		var states = stampit()
			.state(_states);
		var Child = stampit.compose(Validator, states);
		it('should fails rules', function(done) {
            var query2 = {
                action: 'flight'
            };
			var child2 = Child({
				query: query2
			});
			var _query2 = child2.run();
            console.log(Validator.fixed.state.rules.flight); // { ori: 'required|size:3', dst: 'required|size:3' }
            console.log(Child.fixed.state.rules.flight); // { ori: 'size:3', rute: 'required' }
            // if we use deep recursive extend, rules.flight should be { ori: 'size:3', dst: 'required|size:3', rute: 'required' }
			expect(_query2.dst)
				.to.exist; // expecting query.dst to exist because it filled with error message that dst is still required
			done();
		});
```

on console:
``` 
1) Base validator #compose with other states should fails rules:
     AssertionError: expected undefined to exist
```

**Comment by ericelliott** [comment]
I'm not sure I understand the question. Stampit already deep clones state.

```javascript
var b = { b: 'b' };
var foo = stampit().state({ a: b });
var bar = stampit().state({ c: b }); // => { a: { b: 'b' }, c: { b: 'b' } }

example.a.b = 'unique';
example; //=> { a: { b: 'unique' }, c: { b: 'b' } }
```

**Comment by koresar** [comment]
@ericelliott you are wrong. If it does deep cloning then the following code should print '1 2'. But it prints '2 2'.
```js
var stamp = require('stampit')();
var state = { a: { b : null } };

state.a.b = 1;
var obj1 = stamp.create(state);
state.a.b = 2;
var obj2 = stamp.create(state);

console.log(obj1.a.b, obj2.a.b); // 2 2
```
Moreover, [that](https://github.com/ericelliott/stampit/issues/14#issuecomment-24015535) your statement contradicts with my code above.

I first though that it is the issue of the [`mout/object/merge`](https://github.com/ericelliott/stampit/blob/master/stampit.js#L80) function. But actually it is not. See yourself.
```js
var merge = require('mout/object/merge');
var state = { a: { b: null } };

state.a.b = 1;
var obj1 = merge({}, state);

state.a.b = 2;
var obj2 = merge({}, state);

console.log(obj1.a.b, obj2.a.b); // 1 2
```

Let me investigate further.

**Comment by ericelliott** [comment]
Oh yeah, you are right. =) I made a conscious decision not to traverse the entire object hierarchy so that you could pass along things like reference to request / response objects. I think we can open that for debate, but it will be a backward-breaking change, so we'd have to do a 2.0 release if we decide to change it... it's going to effect all the stamps that rely on reference behavior.

**Comment by koresar** [comment]
I found the cause of the unexpected behavior.

The state is deep cloned using [merge function](https://github.com/ericelliott/stampit/blob/master/stampit.js#L80) only when it is part of stamp (i.e. `fixed.state`).
The state is shallow copied using [mixIn function](https://github.com/ericelliott/stampit/blob/master/stampit.js#L81) when it is passed to the stamp function as an instantiation parameter (i.e. `stamp(myState)` or `stamp.create(myState)`).

I propose to make them both deeply cloned. People expect it deeply cloned both times.

**Comment by ericelliott** [comment]
Oh wow. They should not be different! That's a bug.

**Comment by ericelliott** [comment]
@koresar You want to take a stab at a PR?

**Comment by koresar** [comment]
Let's discuss the best approach in the chat https://gitter.im/ericelliott/stampit

**Comment by ericelliott** [comment]
```javascript
var stamp = require('stampit')();
var state = { a: { b : null } };

state.a.b = 1;
var obj1 = stamp.create(state);
state.a.b = 2;
var obj2 = stamp.create(state);

console.log(obj1.a.b, obj2.a.b); // 2 2
```

Regardless of the deep or shallow question, this is definitely a bug.

**Comment by koresar** [comment]
One more nasty bug with the state merging.
```js
var stamp = require('stampit')();
var stateOriginal = { a: { b : 'original', Y: 'original' } };
var stateOverride = { a: { b : 'override', X: 'override' } };

var obj1 = stamp.state(stateOriginal).create();
var obj2 = stamp.state(stateOverride).create();

console.log(obj1.a); // { b: 'original', Y: 'original' } - OK
console.log(obj2.a); // { b: 'override', X: 'override' } - UNEXPECTED
// The expected value : { b: 'override', Y: 'original', X: 'override' }
```

The bug is in [this line](https://github.com/ericelliott/stampit/blob/master/stampit.js#L115). Again, we are using `mixIn` instead of `merge` for state(s).

**Comment by koresar** [comment]
Crap! Similar bug is in the `stampit.compose()` function. Same misuse of the `mixIn`. I reckon this is a copy&paste bug. I'll fix'em all.

**Comment by ericelliott** [comment]
:+1:

**Comment by koresar** [comment]
Will be fixed with #43 in version 2.0. See branch [v2_0](https://github.com/ericelliott/stampit/tree/v2_0).
